README.md
# Blacksad
WIP of my personnal project using LidarLite device.
The idea is to assemble it on a pan&tilt head, then retreive the distance from the LidarLite, the pan and tilt angles from the head.
With this, calculate x,y,z coordinates and stream them on an OSC protocol.
On the reciver part, get this values and open them in Maya or MotionBuilder.
The idea is to be able to create a point cloud in maya with this cheap solution.
Check on the wiki for more infos.


changelog.md
ChangeLog

v0.22.0
  Nothing is written yet in Blacksad program. So Far only test programs a part.
  Stepper.py is working for making actionning my steppers.
  Switches.py is working to read the value of my two switches.

v0.22.1
  Today test : create the driver for LidarLite in python
  Ben en fait ça marche pas .... il faut que je démonte le module ... abandon puis GTA
  :-(

v0.24.1
  Premier Script python qui fait marcher le LidarLite sur le RaspberryPi.
  Récupération de la distance mais juste la petite valeur (0x90) maintenant il me reste a l'implémenter dans le 0x8f.
  


driver_test.py
import smbus
import time

#To get a mesure, we need to write 0x04 in the register 0x00 then read the register 0x8f and transform its result.

bus = smbus.SMBus(1)             # I2C Bus for a RaspberryPi 2
LIDARLite_ADDRESS = 0x62         # I2C address of the LidarLite device
RegisterMeasure = 0x00           # Register to write to initiate ranging.
MeasureValue = 0x04              # Value to initiate ranging.
RegisterHighLowB = 0x8f          # Register to get both High and Low bytes in 1 call.


def I2C_write(register,value):
        bus.write_byte_data(LIDARLite_ADDRESS, register, value)


def I2C_read(register):
        value = bus.read_byte_data(LIDARLite_ADDRESS, register)
        return(value)


try:
    while True:
        I2C_write(RegisterMeasure,MeasureValue)
        time.sleep(0.5)
        read8f = I2C_read(0x90)
        print read8f

except KeyboardInterrupt:

    print "   !!!!Ctrl C detected !!!!! GoodBye"


stepper.py
#!/usr/bin/env python

# Import required libraries
import sys
import time
import RPi.GPIO as GPIO

# Use BCM GPIO references
# instead of physical pin numbers
GPIO.setmode(GPIO.BCM)

# Define GPIO signals to use
# Physical pins on verra ca apres
# GPIO17,GPIO22,GPIO23,GPIO24 pareil a changer
StepPins = [6,13,19,26]

# Set all pins as output
for pin in StepPins:
  print "Setup pins"
  GPIO.setup(pin,GPIO.OUT)
  GPIO.output(pin, False)

# Define advanced sequence
# as shown in manufacturers datasheet
Seq = [[1,0,0,0],
       [1,1,0,0],
       [0,1,0,0],
       [0,1,1,0],
       [0,0,1,0],
       [0,0,1,1],
       [0,0,0,1],
       [1,0,0,1]]

StepCount = len(Seq)-1
StepDir = 2 # Set to 1 or 2 for clockwise
            # Set to -1 or -2 for anti-clockwise

# Read wait time from command line
if len(sys.argv)>1:
  WaitTime = int(sys.argv[1])/float(1000)
else:
  WaitTime = 10/float(1000)

# Initialise variables
StepCounter = 0

# Start main loop
while True:

  for pin in range(0, 4):
    xpin = StepPins[pin]
    print StepCounter
    print pin
    if Seq[StepCounter][pin]!=0:
      print " Step %i Enable %i" %(StepCounter,xpin)
      GPIO.output(xpin, True)
    else:
      GPIO.output(xpin, False)

  StepCounter += StepDir

  # If we reach the end of the sequence
  # start again
  if (StepCounter>=StepCount):
    StepCounter = 0
  if (StepCounter<0):
    StepCounter = StepCount

  # Wait before moving on
  time.sleep(WaitTime)


stepper2.py
#!/usr/bin/env python

# Import required libraries
import sys
import time
import RPi.GPIO as GPIO

# Use BCM GPIO references
# instead of physical pin numbers
GPIO.setmode(GPIO.BCM)

# Define GPIO signals to use
# Physical pins on verra ca apres
# GPIO17,GPIO22,GPIO23,GPIO24 pareil a changer
StepPins = [12,16,20,21]

# Set all pins as output
for pin in StepPins:
  print "Setup pins"
  GPIO.setup(pin,GPIO.OUT)
  GPIO.output(pin, False)

# Define advanced sequence
# as shown in manufacturers datasheet
Seq = [[1,0,0,0],
       [1,1,0,0],
       [0,1,0,0],
       [0,1,1,0],
       [0,0,1,0],
       [0,0,1,1],
       [0,0,0,1],
       [1,0,0,1]]

StepCount = len(Seq)-1
StepDir = 2 # Set to 1 or 2 for clockwise
            # Set to -1 or -2 for anti-clockwise

# Read wait time from command line
if len(sys.argv)>1:
  WaitTime = int(sys.argv[1])/float(1000)
else:
  WaitTime = 10/float(1000)

# Initialise variables
StepCounter = 0

# Start main loop
while True:

  for pin in range(0, 4):
    xpin = StepPins[pin]
    print StepCounter
    print pin
    if Seq[StepCounter][pin]!=0:
      print " Step %i Enable %i" %(StepCounter,xpin)
      GPIO.output(xpin, True)
    else:
      GPIO.output(xpin, False)

  StepCounter += StepDir

  # If we reach the end of the sequence
  # start again
  if (StepCounter>=StepCount):
    StepCounter = 0
  if (StepCounter<0):
    StepCounter = StepCount

  # Wait before moving on
  time.sleep(WaitTime)


Switchs.py
#! /usr/bin/python

import sys
import time
import RPi.GPIO as GPIO
GPIO.setmode(GPIO.BCM) #Using BCM pinout
input_list = [22,23] #GPIO inputs for the switches
output_list = [6,12,13,16,19,20,21,26] #GPIO outputs for the steppers and in the future the laser led
GPIO.setup(input_list, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
GPIO.setup(output_list, GPIO.OUT, initial=GPIO.LOW)
def Sw01action(channel):
    if GPIO.input(22):
        print('This is a edge event callback function!')
        print('Edge detected on channel %s'%channel)
        print('The SW01 is pressed')
    else:
        print('This is a edge event callback function!')
        print('Edge detected on channel %s'%channel)
        print('The SW01 is released')

def Sw02action(channel):
    if GPIO.input(23):
        print('This is a edge event callback function!')
        print('Edge detected on channel %s'%channel)
        print('The SW02 is pressed')
    else:
        print('This is a edge event callback function!')
        print('Edge detected on channel %s'%channel)
        print('The SW02 is released')


GPIO.add_event_detect(22, GPIO.BOTH, callback=Sw01action, bouncetime=200)
GPIO.add_event_detect(23, GPIO.BOTH, callback=Sw02action, bouncetime=200)

try:
    while True:
        pass


except KeyboardInterrupt:
        GPIO.cleanup()
        print "   !!!!Ctrl C detected !!!!! GoodBye"
